jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/utils/JMSUtils.java

```diff
@@ -0,0 +1,180 @@
+                    .put(JMSConstants.PROVIDER_URL, context.getEnvironment().get(JMSConstants.PROVIDER_URL));
+            if (null != context.getEnvironment().get(JMSConstants.CONNECTION_STRING)) {
+                initialContextProperties.put(JMSConstants.CONNECTION_STRING,
+                        context.getEnvironment().get(JMSConstants.CONNECTION_STRING));
+            }
+            if (JMSConstants.DESTINATION_TYPE_TOPIC.equalsIgnoreCase(destinationType)) {
+                initialContextProperties.put(JMSConstants.TOPIC_PREFIX + destinationName, destinationName);
+            } else if (JMSConstants.DESTINATION_TYPE_QUEUE.equalsIgnoreCase(destinationType)
+                    || JMSConstants.DESTINATION_TYPE_GENERIC.equalsIgnoreCase(destinationType)) {
+                initialContextProperties.put(JMSConstants.QUEUE_PREFIX + destinationName, destinationName);
+            }
+            InitialContext initialContext = new InitialContext(initialContextProperties);
+            try {
+                return JMSUtils.lookup(initialContext, Destination.class, destinationName);
+            } catch (NamingException e1) {
+                return JMSUtils.lookup(context, Destination.class,
+                        (JMSConstants.DESTINATION_TYPE_TOPIC.equalsIgnoreCase(destinationType) ?
+                                "dynamicTopics/" :
+                                "dynamicQueues/") + destinationName);
+            }
+
+        }
+    }
+
+    /**
+     * JNDI look up in the context.
+     *
+     * @param context Context that need to looked up
+     * @param clazz   Class of the object that need to be found
+     * @param name    Name of the object that need to be looked up
+     * @param <T>     Class of the Object that need to be found
+     * @return the relevant object, if found in the context
+     * @throws NamingException, if the found object is different from the expected object
+     */
+    private static <T> T lookup(Context context, Class<T> clazz, String name) throws NamingException {
+        Object object = context.lookup(name);
+        try {
+            return clazz.cast(object);
+        } catch (ClassCastException ex) {
+            // Instead of a ClassCastException, throw an exception with some
+            // more information.
+            if (object instanceof Reference) {
+                Reference ref = (Reference) object;
+                String errorMessage =
+                        "JNDI failed to de-reference Reference with name " + name + "; is the " + "factory " + ref
+                                .getFactoryClassName() + " in your classpath?";
+                throw new NamingException(errorMessage);
+            } else {
+                String errorMessage =
+                        "JNDI lookup of name " + name + " returned a " + object.getClass().getName() + " while a "
+                                + clazz + " was expected";
+                logger.error(errorMessage);
```

Comment:
Don't log and throw the error. This is not the best practice, either you handle it or throw it.